src/lustre/lustreTypeChecker.ml

```diff
@@ -318,223 +318,225 @@ and check_type_expr: tc_context -> LA.expr -> tc_type -> unit tc_result
-                                       ^ string_of_tc_type inf_ty))
+    infer_type_expr ctx idx >>= fun index_type -> 
+    if is_expr_int_type ctx idx
+    then infer_type_expr ctx e >>= fun inf_arr_ty ->
+        (match inf_arr_ty with
+          | ArrayType (_, (arr_b_ty, _)) ->
+            R.guard_with(eq_lustre_type ctx arr_b_ty exp_ty)
+              (type_error pos ("Exptected array of base type "
+                                ^ string_of_tc_type exp_ty
+                                ^ " but found type "
+                                ^ string_of_tc_type arr_b_ty))
+          | _ -> type_error pos ("Expected an array type but found type " ^ string_of_tc_type inf_arr_ty))
+    else type_error pos ("Array index should have type int"
+                        ^ " but found type " ^ string_of_tc_type index_type)
+  | (ArraySlice _ as e) | (ArrayConcat _ as e) ->
+    infer_type_expr ctx e >>= fun inf_ty ->
+    R.guard_with(eq_lustre_type ctx inf_ty exp_ty)
+      (type_error (LH.pos_of_expr e) ("Expected type "
+                                      ^ string_of_tc_type exp_ty
+                                      ^ " but found type "
+                                      ^ string_of_tc_type inf_ty))
   (* Quantified expressions *)
   | Quantifier (_, _, qs, e) ->
-     let extn_ctx = List.fold_left union ctx
-                      (List.map (fun (_, i, ty) -> singleton_ty i ty) qs) in
-     check_type_expr extn_ctx e exp_ty
+    let extn_ctx = List.fold_left union ctx
+                    (List.map (fun (_, i, ty) -> singleton_ty i ty) qs) in
+    check_type_expr extn_ctx e exp_ty
 
   (* Clock operators *)
   | When (_, e, _) -> check_type_expr ctx e exp_ty
   | Current (_, e) -> check_type_expr ctx e exp_ty
   | Condact (pos, c, _, node, args, defaults) ->
-     check_type_expr ctx c (Bool pos)
-     >> check_type_expr ctx (Call (pos, node, args)) exp_ty
-     >>  R.seq (List.map (infer_type_expr ctx) defaults)
-     >>= fun d_tys -> R.guard_with (eq_lustre_type ctx exp_ty (GroupType (pos, d_tys)))
-                        (type_error pos "Condact defaults do not have the same type as node call")
+    check_type_expr ctx c (Bool pos)
+    >> check_type_expr ctx (Call (pos, node, args)) exp_ty
+    >>  R.seq (List.map (infer_type_expr ctx) defaults)
+    >>= fun d_tys -> R.guard_with (eq_lustre_type ctx exp_ty (GroupType (pos, d_tys)))
+                      (type_error pos "Condact defaults do not have the same type as node call")
   | Activate (pos, node, cond, _, args) -> 
-     check_type_expr ctx cond (Bool pos)
-     >> check_type_expr ctx (Call (pos, node, args)) exp_ty 
+    check_type_expr ctx cond (Bool pos)
+    >> check_type_expr ctx (Call (pos, node, args)) exp_ty 
   | Merge (pos, i, mcases) ->
-     infer_type_expr ctx (LA.Ident (pos, i))
-     >> R.seq_ (List.map (fun e -> check_type_expr ctx e exp_ty)
-                  (List.map snd mcases))
+    infer_type_expr ctx (LA.Ident (pos, i)) >>= fun ty ->
```

Comment:
I get the following warning: `Warning 27 [unused-var-strict]: unused variable ty.`